Completions do not show for function with same name as mod

JoshMcguigan opened the pull request:
fixes #3444 

I've added a test case in `crates/ra_ide/src/completion/complete_path.rs` which verifies the described behavior in #3444. Digging in, I found that [the module scope iterator](https://github.com/JoshMcguigan/rust-analyzer/blob/ba62d8bd1ce8a68b8d21aaf89ae1ea6787f18366/crates/ra_ide/src/completion/complete_path.rs#L22) only provides the module `z`, and does not provide the function `z` (although if I name the function something else then it does show up here). 

I thought perhaps the name wasn't being properly resolved, but I added a test in `crates/ra_hir_def/src/nameres/tests.rs` which seems to suggest that it is? I've tried to figure out how to bridge the gap between these two tests (one passing, one failing) to see where the function `z` is being dropped, but to this point I haven't been able to track it down.

Any pointers on where I might look for this?

edwin0cheng commented:
@JoshMcguigan The actual resolve is in here, you may add some dbg! to see if there is any unresolve import.  https://github.com/rust-analyzer/rust-analyzer/blob/a6133206d1614624bbb51650d33fae04d70bc21a/crates/ra_hir_def/src/nameres/collector.rs#L284

OTOH, `RawItemsCollector` collect and translate the raw ast item. I think it may be helpful too :

https://github.com/rust-analyzer/rust-analyzer/blob/5e78036e6c8752fda350818afdd411ab25f405ce/crates/ra_hir_def/src/nameres/raw.rs#L226

JoshMcguigan commented:
https://github.com/rust-analyzer/rust-analyzer/blob/5e78036e6c8752fda350818afdd411ab25f405ce/crates/ra_hir_def/src/nameres/raw.rs#L270

I can see the `z` function passing through here. I think that confirms what the test I added in `crates/ra_hir_def/src/nameres/tests.rs` says, which is that it is being collected?

edwin0cheng commented:
I think so ? The next step would be check if it appears in :

https://github.com/rust-analyzer/rust-analyzer/blob/a6133206d1614624bbb51650d33fae04d70bc21a/crates/ra_hir_def/src/nameres/collector.rs#L765

And does it update the `PerNs` ?

https://github.com/rust-analyzer/rust-analyzer/blob/a6133206d1614624bbb51650d33fae04d70bc21a/crates/ra_hir_def/src/nameres/collector.rs#L471

matklad commented:
@JoshMcguigan you are correct that the function *should* be present in the `module_scope`. And indeed, the most likely culprit is the code inside `DefCollector`, `ModCollector`. It's ... a rather complicated bit of code which ideally needs to be rewritten in a more principled fashion.

JoshMcguigan commented:
```rust
// crates/ra_hir_def/src/nameres/tests.rs
#[test]
fn crate_def_map_fn_mod_same_name() {
    let map = def_map(
        "
        //- /lib.rs
        mod m {
            pub mod z {}
            pub fn z() {}
        }
        ",
    );
    assert_snapshot!(map, @r###"
        ⋮crate
        ⋮m: t
        ⋮
        ⋮crate::m
        ⋮z: t v
        ⋮
        ⋮crate::m::z
    "###)
}
```

Just to clarify, this test passes. Doesn't that mean that the collector is doing the right thing? If not, is there a different test I could add to cover this behaviour of the collector?

matklad commented:
@JoshMcguigan this means that *some* part of collector works OK. But, IIRC, the code that leads to `z: t v` and the code than enumerates `module_scope` for completion uses the same data structure from `DefCollector`, so it looks like DefCollector misbihaves in the real case.

JoshMcguigan commented:
I think I've tracked the problem down to [this conversion from PerNs to ScopeDef](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_hir/src/code_model.rs#L229). I've confirmed that the `PerNs` for `z` has both the function and the module, but it seems the [from<PerNs> for ScopeDef impl](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_hir/src/code_model.rs#L1291) will only return one of them (favoring the module since it takes types before values).

matklad commented:
Oh, right.... 

This means that the fix needs to be somewhat creative: we need to add items from all namespaces, but only if they are different. 

That is, `struct Foo;` should be present only once (atthough it occupies both types and values namespace) but for module and function with the same name, we should show both .

Sorry for leading you onto the wrong track @JoshMcguigan !

JoshMcguigan commented:
Alright, I've taken a very rough initial pass at implementing a fix for this. Basically it creates a new method to convert from `PerNs` to `ScopeDef` which allows returning more than one item. 

Does this seems to be on the right track? I'd guess we'd want to remove the `impl From<PerNs> for ScopeDef` and switch over to using this new method in all places that convert from a `PerNs` to a `ScopeDef`?

matklad commented on crates/ra_hir/src/code_model.rs:
```diff
@@ -1300,6 +1304,38 @@ impl From<PerNs> for ScopeDef {
 }
 
+impl ScopeDef {
+    fn all_items(def: PerNs) -> Vec<Self> {
```
Let's either return an `ArrayVec<[ScopeDef; 2]>` here, or maybe just accept an `acc: &mut Vec<ScopeDef>` as an argument, to avoid the needless allocation.

bjorn3 replied:
I think it can return 3 results: a type, a value and a macro.

matklad replied:
Edited the comment!

bjorn3 replied:
Where did you define the function `O` in that snippet? :)

JoshMcguigan replied:
I replaced this `ArrayVec<[Self; 3]>`. Let me know if that isn't what you were looking for.

JoshMcguigan commented:
I think at this point this is potentially ready to merge. I removed the `From<PerNs>` impl, and replaced it with a call to the new `all_items` method. 

The one thing that sticks out to me as less than ideal is

```rust
// code_model.rs
if items.is_empty() {
    items.push(ScopeDef::Unknown)
}
```

This feels like it should be the responsibility/decision of the code calling this function. Deleting this only makes a single test fail, `dont_complete_current_use` in `complete_path.rs`, and it seems to only fail because of the `covers!(dont_complete_current_use);`. 

Removing the `covers!` from the test makes it pass, but then I wasn't sure about how to test (or if we should remove) the logic that test was supposed to be covering. 

Is this okay as-is? Or is this something that can/should be cleaned up in a future PR? Or is this something worth doing in this PR?

matklad commented:
>This feels like it should be the responsibility/decision of the code calling this function. Deleting this only makes a single test fail, dont_complete_current_use in complete_path.rs, and it seems to only fail because of the covers!(dont_complete_current_use);.

Wow, this I think is the first documented case of test marks preventing code from suddenly and silently becoming dead!

I've added a test that should be there from the start: https://github.com/rust-analyzer/rust-analyzer/pull/3572/files. The idea is that we want to complete even unresolved use paths. So `ScopeDef::Unknown` means "there's this name in scope, but it don't know what this name actually means".

matklad commented:
bors r+

Thanks @JoshMcguigan !
